app/src/main/java/org/connectbot/TerminalView.java

```diff
@@ -150,6 +170,282 @@ public TerminalView(Context context, TerminalBridge bridge) {
 
 		// Enable accessibility features if a screen reader is active.
 		new AccessibilityStateTester().execute((Void) null);
+
+		clipboard = (ClipboardManager) context.getSystemService(Context.CLIPBOARD_SERVICE);
+
+		setTextColor(0x00000000);
+		setTypeface(Typeface.MONOSPACE);
+		onFontSizeChanged(bridge.getFontSize());
+
+		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB) {
+			setTextIsSelectable(true);
+
+			initSelectionCallback();
+
+			gestureDetector = new GestureDetector(context, new GestureDetector.SimpleOnGestureListener() {
+				private TerminalBridge bridge = TerminalView.this.bridge;
+				private float totalY = 0;
+
+				@Override
+				public boolean onScroll(MotionEvent e1, MotionEvent e2, float distanceX, float distanceY) {
+					// if releasing then reset total scroll
+					if (e2.getAction() == MotionEvent.ACTION_UP) {
+						totalY = 0;
+					}
+
+					totalY += distanceY;
+					final int moved = (int) (totalY / bridge.charHeight);
+
+					if (moved != 0) {
+						int base = bridge.buffer.getWindowBase();
+						bridge.buffer.setWindowBase(base + moved);
+						totalY = 0;
+
+						copyBufferToText();
+					}
+
+					return true;
+				}
+
+				@Override
+				public boolean onSingleTapConfirmed(MotionEvent e) {
+					viewPager.performClick();
+					return super.onSingleTapConfirmed(e);
+				}
+			});
+		}
+	}
+
+	@TargetApi(11)
+	private void initSelectionCallback() {
+		this.setCustomSelectionActionModeCallback(new ActionMode.Callback() {
+			private static final int PASTE = 0;
+
+			@Override
+			public boolean onCreateActionMode(ActionMode mode, Menu menu) {
+				TerminalView.this.selectionActionMode = mode;
+
+				menu.add(0, PASTE, 2, "Paste")
+						.setIcon(R.drawable.ic_action_paste)
+						.setShowAsAction(MenuItem.SHOW_AS_ACTION_WITH_TEXT | MenuItem.SHOW_AS_ACTION_IF_ROOM);
+
+				return true;
+			}
+
+			@Override
+			public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
+				if (item.getItemId() == PASTE) {
+					String clip = clipboard.getText().toString();
+					TerminalView.this.bridge.injectString(clip);
+					mode.finish();
+					return true;
+				}
+
+				return false;
+			}
+
+			@Override
+			public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
+				return false;
+			}
+
+			@Override
+			public void onDestroyActionMode(ActionMode mode) {
+			}
+		});
+	}
+
+	public void copyCurrentSelectionToClipboard() {
+		ClipboardManager clipboard =
+				(ClipboardManager) TerminalView.this.context.getSystemService(Context.CLIPBOARD_SERVICE);
+		if (currentSelection.length() != 0) {
+			clipboard.setText(currentSelection);
+		}
+
+		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB &&
+			selectionActionMode != null) {
+			selectionActionMode.finish();
+			selectionActionMode = null;
+		}
+	}
+
+	@Override
+	protected void onSelectionChanged(int selStart, int selEnd) {
+		if (selStart <= selEnd) {
```

Comment:
TerminalView maintains a currentSelection, which is used when the user presses Ctrl+Shift+V on hard keyboard. This is why onSelectionChanged is overridden.

Unfortunately sometimes quickly moving the mouse selection results in `selEnd` being less than `selStart`, crashing the app without this check when it his `substring()`.